### ISSUE TITLE
Maven build for Bigtable

### ISSUE THREAD

**Issue by dodizzle** [JanusGraph/janusgraph#559]
I am following the instructions http://docs.janusgraph.org/0.2.0-SNAPSHOT/bigtable.html
I have edited the pom file at the root of the project and added the 2 dependancies.
I can build and create the docker image but the bigtable-hbase-1.x-hadoop and netty-tcnative-boringssl-static jar files do not exist in the image.

find /var/ | grep bigtable comes up empty. 
And when the container is running ps auxww| grep bigtable also comes up empty.

Am I not adding the dependancies to the correct pom file?

@mbrukman

**Comment by dodizzle** [comment]
Upate: Apparently the command I am using to build the docker image that only supports elasticsearch and hadoop.
The command is documented here:
https://github.com/JanusGraph/janusgraph/blob/master/BUILDING.md
Which is mvn clean install -Pjanusgraph-release -Dgpg.skip=true -DskipTests=true && mvn docker:build -Pjanusgraph-docker -pl janusgraph-dist

How do I build for hbase/big table instead?

**Comment by igorbernstein2** [comment]
I'm not too familiar with with janusgraph's build system, so I'm not sure if this is the correct way to include bigtable dependencies, but I think this should work:

edit janusgraph-all/pom.xml and add the following to the <dependencies> section:
```xml
<dependency>
    <groupId>com.google.cloud.bigtable</groupId>
    <artifactId>bigtable-hbase-1.x-hadoop</artifactId>
    <version>1.0.0-pre3</version>
</dependency>

<dependency>
   <groupId>io.netty</groupId>
   <artifactId>netty-tcnative-boringssl-static</artifactId>
   <version>1.1.33.Fork26</version>
</dependency>
```

Then run your build command:
`mvn clean install -Pjanusgraph-release -Dgpg.skip=true -DskipTests=true && mvn docker:build -Pjanusgraph-docker -pl janusgraph-dist`

**Comment by sjudeng** [comment]
Also currently the referenced Docker build is specifically for use with the BerkeleyJE storage backend and Elasticsearch indexing backend. In your case I think you'd need to setup and manage your own HBase/Big Table and Elasticsearch instances (using Docker or otherwise) and then connect through Gremlin shell or Gremlin server using the JanusGraph distribution, with Bigtable dependencies added to your classpath as relevant. Some of the docs linked below are hopefully helpful on this.

1. https://cloud.google.com/bigtable/docs/installing-hbase-shell
1. https://www.elastic.co/guide/en/elasticsearch/reference/current/_installation.html
1. http://docs.janusgraph.org/latest/getting-started.html#_downloading_janusgraph_and_running_the_gremlin_shell
1. http://docs.janusgraph.org/latest/server.html

**Comment by dodizzle** [comment]
I have made some progress.
I added 2 jar files as described here  http://docs.janusgraph.org/0.2.0-SNAPSHOT/bigtable.html to the classpath which is great.
Now is it attempting to connect to the big table instance that I described in bigtable.properties but failing because of a permissions problem.
My guess is there is additional hbase specific configs that are missing such as tablename etc...
Some of the errors I am seeing are:

java.lang.RuntimeException: org.janusgraph.diskstorage.TemporaryBackendException: Temporary failure in storage backend
Caused by: java.io.IOException: Failed to listTables
Caused by: com.google.bigtable.repackaged.io.grpc.StatusRuntimeException: PERMISSION_DENIED: Request had insufficient authentication scopes.

BTW within this same container I can write to the bigtable instance using cbt and the same bigtable settings.

**Comment by igorbernstein2** [comment]
You need to add credentials into your container and configure the bigtable connector to use them:

1. Decide if you want to:
-  use you personal application default credentials that are created via the gcloud sdk (https://developers.google.com/identity/protocols/application-default-credentials)
- use a service account (https://cloud.google.com/iam/docs/understanding-service-accounts)
2. Mount the json file as volume (https://docs.docker.com/engine/admin/volumes/volumes/#start-a-container-with-a-volume)
(https://docs.docker.com/engine/reference/run/#env-environment-variables)
3. Configure the bigtable connector:
- set the GOOGLE_APPLICATION_CREDENTIALS environment variable to the location of the json file inside your container
- or update your janus graph config to set the google.bigtable.auth.json.keyfile to the location of the mounted json file

**Comment by dodizzle** [comment]
I got this to work. Thanks for your help.
Should I document what I did to get this working?

**Comment by mbrukman** [comment]
@dodizzle — yes, please! Documentation updates are always appreciated; please send a PR to update [`docs/bigtable.txt`][1] (note that it's in Asciidoc format). Before we can accept your contribution, we need you to [sign the JanusGraph CLA][2].

[1]: https://github.com/JanusGraph/janusgraph/blob/master/docs/bigtable.txt
[2]: https://github.com/JanusGraph/janusgraph/blob/master/CONTRIBUTING.md

**Comment by mbrukman** [comment]
@dodizzle — please also see if my changes in PR #592 work for you as well (basically, this is what @igorbernstein2 suggested above), and leave a comment on that PR either way. Thanks!

**Comment by mbrukman** [comment]
@dodizzle — PR #592 was merged, so now you can just pull the latest 0.2.0-SNAPSHOT from `master` branch and rebuild JanusGraph, and let us know if it works for you. Let us know if we should make any changes before this becomes the 0.2.0 release!

**Comment by mbrukman** [comment]
The 0.2.0 release is imminent, so we won't be able to make any changes or tweaks to this integration before it's out.

I'm going to close this issue as we have addressed the original report by directly integrating the Cloud Bigtable client into the Maven build for JanusGraph so the original instructions for modifying the `pom.xml` file are no longer needed — docs have been updated accordingly. Now, you just need to configure the HBase backend to connect to Bigtable, so it's just a few lines of config, no rebuilding needed.

If there are any further questions or comments, please send email to the developer list and we can discuss.

If there's a specific problem or bug, please open a new issue with a specific description of what you would like to see addressed, so that each issue is about a single, specific task to be addressed.

Thanks!
